keanu-project/src/main/java/io/improbable/keanu/vertices/VertexId.java

```diff
@@ -0,0 +1,64 @@
+package io.improbable.keanu.vertices;
+
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.google.common.primitives.Ints;
+
+/**
+ * An object representing the ID value of a vertex.  IDs are assigned in such a way that a Lexicographic ordering of
+ * all nodes based on their ID value is a valid topological ordering of the graph made up by those vertices.
+ *
+ * Ids also encapsulate the notion of "Depth".  When we have graphs within graphs, the depth tells us at what level the
+ * graph exists - ie depth 1 is the outermost graph, depth 2 is a graph within a graph etc.
+ */
+public class VertexId implements Comparable<VertexId> {
+
+    public static final AtomicLong ID_GENERATOR = new AtomicLong(0L);
+
+    LinkedList<Long> idValues = new LinkedList<>();
+
+    public VertexId() {
+        long newId = ID_GENERATOR.getAndIncrement();
+        idValues.push(newId);
+    }
+
+    @Override
+    public int compareTo(VertexId that) {
+        long comparisonValue = 0;
+        int  minOfListSizes = Math.min(this.idValues.size(), that.idValues.size());
+
+        for (int i = 0; i < minOfListSizes && comparisonValue == 0; i++) {
+            comparisonValue = this.idValues.get(i) - that.idValues.get(i);
```

Comment:
Using `.get(i)` on a LinkedList is going to be potentially very inefficient. For the compareTo contract I *think* you only need to return a -1, 0 or 1 for the comparison. In which case you should be able to return on id size differences or in the case where the size is the same, the comparison results of the first different id.

Reply:
Yeah - I knew about the .get() being inefficient - as we discussed the LinkedList was a temporary placeholder whilst I went through the standard library to make sure there was nothing else more appropriate.  In my latest code I use an array as we discussed.